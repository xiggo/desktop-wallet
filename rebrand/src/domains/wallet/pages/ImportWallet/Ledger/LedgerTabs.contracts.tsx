import { Networks } from "@payvo/sdk";
import { Contracts } from "@payvo/sdk-profiles";

import { useLedgerScanner } from "@/app/contexts";

export enum LedgerTabStep {
	NetworkStep = 1,
	LedgerConnectionStep,
	LedgerScanStep,
	LedgerImportStep,
}

export interface LedgerTabsProperties {
	activeIndex?: LedgerTabStep;
	onClickEditWalletName: (wallet: Contracts.IReadWriteWallet) => void;
}

export interface LedgerTableProperties extends ReturnType<typeof useLedgerScanner> {
	network: Networks.Network;
}
