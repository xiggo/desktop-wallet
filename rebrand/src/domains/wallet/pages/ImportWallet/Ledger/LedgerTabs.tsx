import { Enums } from "@payvo/sdk";
import { uniq } from "@payvo/sdk-helpers";
import React, { useCallback, useRef, useState } from "react";
import { useFormContext } from "react-hook-form";
import { useTranslation } from "react-i18next";
import { useHistory } from "react-router-dom";

import { LedgerConnectionStep } from "./LedgerConnectionStep";
import { LedgerImportStep } from "./LedgerImportStep";
import { LedgerScanStep } from "./LedgerScanStep";
import { LedgerTabsProperties, LedgerTabStep } from "./LedgerTabs.contracts";
import { Button } from "@/app/components/Button";
import { Icon } from "@/app/components/Icon";
import { StepIndicator } from "@/app/components/StepIndicator";
import { TabPanel, Tabs } from "@/app/components/Tabs";
import { LedgerData, useLedgerContext } from "@/app/contexts";
import { useActiveProfile } from "@/app/hooks";
import { useKeydown } from "@/app/hooks/use-keydown";
import { useWalletConfig } from "@/domains/wallet/hooks";
import { NetworkStep } from "@/domains/wallet/components/NetworkStep";
import { assertWallet } from "@/utils/assertions";

const Paginator = ({
	activeIndex,
	isMultiple,
	isNextDisabled,
	isNextLoading,
	onBack,
	onFinish,
	onNext,
	onRetry,
	showRetry,
	size,
}: {
	activeIndex: LedgerTabStep;
	isMultiple: boolean;
	isNextDisabled?: boolean;
	isNextLoading?: boolean;
	onBack: () => void;
	onFinish: () => void;
	onNext: () => void;
	onRetry?: () => void;
	showRetry?: boolean;
	size: number;
}) => {
	const { t } = useTranslation();

	return (
		<div className="flex justify-between mt-10">
			<div>
				{showRetry && (
					<Button variant="secondary" onClick={onRetry} data-testid="Paginator__retry-button">
						<Icon name="ArrowRotateRight" />
						<span>{t("COMMON.RETRY")}</span>
					</Button>
				)}
			</div>

			<div className="flex space-x-3">
				{activeIndex < size && (
					<Button variant="secondary" onClick={onBack} data-testid="Paginator__back-button">
						{t("COMMON.BACK")}
					</Button>
				)}

				{activeIndex < size && (
					<Button
						disabled={isNextDisabled || isNextLoading}
						isLoading={isNextLoading}
						onClick={onNext}
						data-testid="Paginator__continue-button"
					>
						{t("COMMON.CONTINUE")}
					</Button>
				)}

				{activeIndex === size && (
					<Button disabled={isNextDisabled} data-testid="Paginator__finish-button" onClick={onFinish}>
						{isMultiple ? t("COMMON.GO_TO_PORTFOLIO") : t("COMMON.GO_TO_WALLET")}
					</Button>
				)}
			</div>
		</div>
	);
};

export const LedgerTabs = ({
	activeIndex = LedgerTabStep.NetworkStep,
	onClickEditWalletName,
}: LedgerTabsProperties) => {
	const activeProfile = useActiveProfile();

	const history = useHistory();
	const { importLedgerWallets, isBusy } = useLedgerContext();
	const { selectedNetworkIds, setValue } = useWalletConfig({ profile: activeProfile });

	const { t } = useTranslation();

	const { formState, getValues, handleSubmit } = useFormContext();
	const { isValid, isSubmitting } = formState;

	const [importedWallets, setImportedWallets] = useState<LedgerData[]>([]);
	const [activeTab, setActiveTab] = useState<number>(activeIndex);

	const isMultiple = importedWallets.length > 1;

	const [showRetry, setShowRetry] = useState(false);
	const retryFunctionReference = useRef<() => void>();

	const importWallets = useCallback(
		async ({ network, wallets }: any) => {
			setImportedWallets(wallets);
			const coin = activeProfile.coins().set(network.coin(), network.id());
			await importLedgerWallets(wallets, coin, activeProfile);

			setValue("selectedNetworkIds", uniq([...selectedNetworkIds, coin.network().id()]));
		},
		[importLedgerWallets, activeProfile, setValue, selectedNetworkIds],
	);

	const isNextDisabled = isBusy || !isValid;

	useKeydown("Enter", () => {
		const isButton = (document.activeElement as any)?.type === "button";

		if (!isButton && !isNextDisabled && !isSubmitting) {
			if (activeTab < LedgerTabStep.LedgerImportStep) {
				handleNext();
			} else {
				handleFinish();
			}
		}
	});

	const handleNext = useCallback(async () => {
		if (activeTab === LedgerTabStep.LedgerScanStep) {
			await handleSubmit((data: any) => importWallets(data))();
		}

		setActiveTab(activeTab + 1);
	}, [activeTab, handleSubmit, importWallets]);

	const handleBack = () => {
		if (activeTab === LedgerTabStep.NetworkStep) {
			return history.push(`/profiles/${activeProfile.id()}/dashboard`);
		}

		if (activeTab === LedgerTabStep.LedgerScanStep) {
			return setActiveTab(LedgerTabStep.NetworkStep);
		}

		setActiveTab(activeTab - 1);
	};

	const handleRetry = useCallback(
		(callback?: () => void) => {
			retryFunctionReference.current = callback;
			setShowRetry(!!callback);
		},
		[retryFunctionReference, setShowRetry],
	);

	const handleFinish = useCallback(() => {
		if (isMultiple) {
			history.push(`/profiles/${activeProfile.id()}/dashboard`);
			return;
		}

		const importedWallet = activeProfile
			.wallets()
			.findByAddressWithNetwork(importedWallets[0].address, getValues("network").id());

		assertWallet(importedWallet);
		history.push(`/profiles/${activeProfile.id()}/wallets/${importedWallet.id()}`);
	}, [isMultiple, history, activeProfile, getValues, importedWallets]);

	return (
		<Tabs activeId={activeTab}>
			<StepIndicator size={4} activeIndex={activeTab} />

			<div data-testid="LedgerTabs" className="mt-8">
				<TabPanel tabId={LedgerTabStep.NetworkStep}>
					<NetworkStep
						filter={(network) =>
							network.allows(Enums.FeatureFlag.TransactionTransferLedgerS) ||
							network.allows(Enums.FeatureFlag.TransactionTransferLedgerX)
						}
						profile={activeProfile}
						title={t("WALLETS.PAGE_IMPORT_WALLET.NETWORK_STEP.TITLE")}
						subtitle={t("WALLETS.PAGE_IMPORT_WALLET.NETWORK_STEP.SUBTITLE")}
					/>
				</TabPanel>
				<TabPanel tabId={LedgerTabStep.LedgerConnectionStep}>
					<LedgerConnectionStep onConnect={() => setActiveTab(LedgerTabStep.LedgerScanStep)} />
				</TabPanel>
				<TabPanel tabId={LedgerTabStep.LedgerScanStep}>
					<LedgerScanStep profile={activeProfile} setRetryFn={handleRetry} />
				</TabPanel>
				<TabPanel tabId={LedgerTabStep.LedgerImportStep}>
					<LedgerImportStep
						wallets={importedWallets}
						profile={activeProfile}
						onClickEditWalletName={onClickEditWalletName}
					/>
				</TabPanel>
			</div>

			<Paginator
				activeIndex={activeTab}
				isMultiple={isMultiple}
				isNextDisabled={isNextDisabled}
				isNextLoading={isSubmitting}
				onBack={handleBack}
				onFinish={handleFinish}
				onNext={handleNext}
				onRetry={retryFunctionReference.current}
				showRetry={showRetry}
				size={4}
			/>
		</Tabs>
	);
};
