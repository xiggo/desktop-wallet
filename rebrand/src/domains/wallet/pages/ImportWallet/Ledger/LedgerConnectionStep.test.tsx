import Transport from "@ledgerhq/hw-transport";
import { Contracts } from "@payvo/sdk-profiles";
import { renderHook } from "@testing-library/react-hooks";
import { createMemoryHistory } from "history";
import React from "react";
import { act } from "react-dom/test-utils";
import { FormProvider, useForm } from "react-hook-form";
import { useTranslation } from "react-i18next";
import { Route } from "react-router-dom";

import { LedgerConnectionStep } from "./LedgerConnectionStep";
import { EnvironmentProvider, minVersionList } from "@/app/contexts";
import { LedgerProvider } from "@/app/contexts/Ledger/Ledger";
import {
	env,
	getDefaultLedgerTransport,
	getDefaultProfileId,
	ledgerObserverSpy,
	render,
	screen,
	waitFor,
} from "@/utils/testing-library";

const history = createMemoryHistory();

describe("LedgerConnectionStep", () => {
	let profile: Contracts.IProfile;
	let wallet: Contracts.IReadWriteWallet;
	let transport: typeof Transport;
	let getVersionSpy: jest.SpyInstance;

	beforeAll(async () => {
		profile = env.profiles().findById(getDefaultProfileId());
		await env.profiles().restore(profile);
		await profile.sync();

		wallet = profile.wallets().first();
		getVersionSpy = jest
			.spyOn(wallet.coin().ledger(), "getVersion")
			.mockResolvedValue(minVersionList[wallet.network().coin()]);

		transport = getDefaultLedgerTransport();
	});

	afterAll(() => {
		getVersionSpy.mockRestore();
	});

	const Component = ({ onConnect = jest.fn(), onFailed = jest.fn() }) => {
		const form = useForm({
			defaultValues: {
				network: wallet.network(),
			},
		});

		return (
			<EnvironmentProvider env={env}>
				<FormProvider {...form}>
					<LedgerProvider transport={transport}>
						<LedgerConnectionStep onConnect={onConnect} onFailed={onFailed} />
					</LedgerProvider>
				</FormProvider>
			</EnvironmentProvider>
		);
	};

	it("should emit event on connect", async () => {
		const { result } = renderHook(() => useTranslation());
		const { t } = result.current;

		const publicKeyPaths = new Map([
			["m/44'/111'/0'/0/0", "027716e659220085e41389efc7cf6a05f7f7c659cf3db9126caabce6cda9156582"],
			["m/44'/111'/1'/0/0", wallet.publicKey()!],
			["m/44'/111'/2'/0/0", "020aac4ec02d47d306b394b79d3351c56c1253cd67fe2c1a38ceba59b896d584d1"],
		]);

		const getPublicKeySpy = jest
			.spyOn(wallet.coin().ledger(), "getPublicKey")
			.mockResolvedValue(publicKeyPaths.values().next().value);

		const onConnect = jest.fn();

		history.push(`/profiles/${profile.id()}`);

		const { container } = render(
			<Route path="/profiles/:profileId">
				<Component onConnect={onConnect} />
			</Route>,
			{ history, withProviders: false },
		);

		await expect(screen.findByText(t("WALLETS.MODAL_LEDGER_WALLET.CONNECT_SUCCESS"))).resolves.toBeVisible();

		await waitFor(() => expect(onConnect).toHaveBeenCalledWith());

		expect(container).toMatchSnapshot();

		getPublicKeySpy.mockReset();
	});

	it("should emit event on fail", async () => {
		const { result } = renderHook(() => useTranslation());
		const { t } = result.current;

		const getPublicKeySpy = jest.spyOn(wallet.coin().ledger(), "getPublicKey").mockImplementation(() => {
			throw new Error(t("WALLETS.MODAL_LEDGER_WALLET.GENERIC_CONNECTION_ERROR"));
		});

		const onFailed = jest.fn();

		history.push(`/profiles/${profile.id()}`);

		const { mockTransportListen, observer } = ledgerObserverSpy();
		const transportListenMock = mockTransportListen(transport);

		const { container } = render(
			<Route path="/profiles/:profileId">
				<Component onFailed={onFailed} />
			</Route>,
			{ history, withProviders: false },
		);

		act(() => {
			observer.error(new Error(t("WALLETS.MODAL_LEDGER_WALLET.GENERIC_CONNECTION_ERROR")));
		});

		await expect(
			screen.findByText(t("WALLETS.MODAL_LEDGER_WALLET.GENERIC_CONNECTION_ERROR")),
		).resolves.toBeVisible();

		expect(container).toMatchSnapshot();

		getPublicKeySpy.mockReset();
		transportListenMock.mockRestore();
	});

	it("should show update error if app version is less than minimum version", async () => {
		const { result } = renderHook(() => useTranslation());
		const { t } = result.current;

		const publicKeyPaths = new Map([
			["m/44'/111'/0'/0/0", "027716e659220085e41389efc7cf6a05f7f7c659cf3db9126caabce6cda9156582"],
			["m/44'/111'/1'/0/0", wallet.publicKey()!],
			["m/44'/111'/2'/0/0", "020aac4ec02d47d306b394b79d3351c56c1253cd67fe2c1a38ceba59b896d584d1"],
		]);

		const getPublicKeySpy = jest
			.spyOn(wallet.coin().ledger(), "getPublicKey")
			.mockResolvedValue(publicKeyPaths.values().next().value);

		const outdatedVersion = "1.0.1";
		const getVersionSpy = jest.spyOn(wallet.coin().ledger(), "getVersion").mockResolvedValue(outdatedVersion);

		const onFailed = jest.fn();

		history.push(`/profiles/${profile.id()}`);

		const { container } = render(
			<Route path="/profiles/:profileId">
				<Component onFailed={onFailed} />
			</Route>,
			{ history, withProviders: false },
		);

		expect(
			screen.getByText(
				t("WALLETS.MODAL_LEDGER_WALLET.OPEN_APP", {
					coin: wallet.network().coin(),
				}),
			),
		).toBeInTheDocument();

		await waitFor(() => expect(onFailed).toHaveBeenCalledWith(expect.any(Error)));

		await expect(
			screen.findByText(
				t("WALLETS.MODAL_LEDGER_WALLET.UPDATE_ERROR", {
					coin: wallet.network().coin(),
					version: outdatedVersion,
				}),
			),
		).resolves.toBeVisible();

		expect(container).toMatchSnapshot();

		getPublicKeySpy.mockReset();
		getVersionSpy.mockReset();
	});
});
