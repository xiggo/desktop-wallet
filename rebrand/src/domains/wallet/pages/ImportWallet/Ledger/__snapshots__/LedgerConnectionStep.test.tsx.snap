// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LedgerConnectionStep should emit event on connect 1`] = `
<div>
  <section
    class="space-y-8"
    data-testid="LedgerConnectionStep"
  >
    <div
      class="flex justify-between items-end bg-theme-background"
    >
      <div
        class="space-y-4"
      >
        <h1
          class="mb-0 text-2xl"
          data-testid="header__title"
        >
          Open App on Ledger
        </h1>
        <div
          class="text-theme-secondary-text"
          data-testid="header__subtitle"
        >
          Open the app on your Ledger and check for details.
        </div>
      </div>
    </div>
    <fieldset
      class="FormField__FormFieldStyled-sc-q4bblx-0 dFNJMk flex flex-col space-y-2"
    >
      <label
        class="flex mb-2 text-sm font-semibold transition-colors duration-100 FormLabel text-theme-secondary-text"
        data-testid="FormLabel"
        for="network"
      >
        Cryptoasset
      </label>
      <div>
        <div
          aria-expanded="false"
          aria-haspopup="listbox"
          aria-owns="ImportWallet__network-menu"
          data-testid="SelectNetwork"
          role="combobox"
        >
          <label
            for="ImportWallet__network-input"
            id="ImportWallet__network-label"
          />
          
          <div
            class="Input__InputWrapperStyled-sc-1i8sd1r-0 ktoJbG"
            disabled=""
          >
            <div
              aria-label="ARK Devnet"
              class="Circle__CircleWrapper-sc-d5u41q-0 bcplCx border-theme-secondary-200 dark:border-theme-secondary-700 text-theme-secondary-500 ring-theme-background"
              data-testid="SelectNetworkInput__network"
            >
              <div
                class="sc-bdvvtL qGXhc"
                data-testid="NetworkIcon__icon"
                height="20"
                width="20"
              >
                <svg>
                  ark.svg
                </svg>
              </div>
            </div>
            <div
              class="relative flex flex-1 h-full"
            >
              <input
                aria-autocomplete="list"
                aria-controls="ImportWallet__network-menu"
                aria-labelledby="ImportWallet__network-label"
                autocomplete="off"
                class="Input__InputStyled-sc-1i8sd1r-1 hPntwa p-0 border-none bg-transparent focus:ring-0 no-ligatures w-full text-theme-secondary-text"
                data-testid="SelectNetworkInput__input"
                disabled=""
                id="ImportWallet__network-input"
                placeholder="Enter a cryptoasset name"
                value="ARK Devnet"
              />
              
            </div>
          </div>
        </div>
        <div
          class=""
          data-testid="SelectNetwork__options"
        >
          <div
            class="mt-6 hidden"
          >
            <ul
              aria-labelledby="ImportWallet__network-label"
              class="grid grid-cols-7 gap-3 mt-3"
              id="ImportWallet__network-menu"
              role="listbox"
            />
          </div>
          <div
            class="mt-6 hidden"
          >
            <ul
              aria-labelledby="ImportWallet__network-label"
              class="grid grid-cols-7 gap-3 mt-3"
              id="ImportWallet__network-menu"
              role="listbox"
            />
          </div>
        </div>
      </div>
    </fieldset>
    <div
      class="flex overflow-hidden flex-col space-y-4 rounded-lg p-4 bg-theme-success-50 dark:bg-theme-secondary-800 sm:space-y-0 sm:space-x-4 sm:flex-row sm:items-center"
    >
      <div
        class="h-11 w-11 flex flex-shrink-0 items-center justify-center rounded-lg text-white bg-theme-success-600"
      >
        <div
          class="sc-bdvvtL qGXhc"
          height="20"
          width="20"
        >
          <svg>
            circle-check-mark.svg
          </svg>
        </div>
      </div>
      <div
        class="flex flex-col w-full"
      >
        <div
          class="text-sm leading-relaxed break-words dark:text-theme-secondary-500"
        >
          Successfully connected
        </div>
      </div>
    </div>
  </section>
</div>
`;

exports[`LedgerConnectionStep should emit event on fail 1`] = `
<div>
  <section
    class="space-y-8"
    data-testid="LedgerConnectionStep"
  >
    <div
      class="flex justify-between items-end bg-theme-background"
    >
      <div
        class="space-y-4"
      >
        <h1
          class="mb-0 text-2xl"
          data-testid="header__title"
        >
          Open App on Ledger
        </h1>
        <div
          class="text-theme-secondary-text"
          data-testid="header__subtitle"
        >
          Open the app on your Ledger and check for details.
        </div>
      </div>
    </div>
    <fieldset
      class="FormField__FormFieldStyled-sc-q4bblx-0 dFNJMk flex flex-col space-y-2"
    >
      <label
        class="flex mb-2 text-sm font-semibold transition-colors duration-100 FormLabel text-theme-secondary-text"
        data-testid="FormLabel"
        for="network"
      >
        Cryptoasset
      </label>
      <div>
        <div
          aria-expanded="false"
          aria-haspopup="listbox"
          aria-owns="ImportWallet__network-menu"
          data-testid="SelectNetwork"
          role="combobox"
        >
          <label
            for="ImportWallet__network-input"
            id="ImportWallet__network-label"
          />
          
          <div
            class="Input__InputWrapperStyled-sc-1i8sd1r-0 ktoJbG"
            disabled=""
          >
            <div
              aria-label="ARK Devnet"
              class="Circle__CircleWrapper-sc-d5u41q-0 bcplCx border-theme-secondary-200 dark:border-theme-secondary-700 text-theme-secondary-500 ring-theme-background"
              data-testid="SelectNetworkInput__network"
            >
              <div
                class="sc-bdvvtL qGXhc"
                data-testid="NetworkIcon__icon"
                height="20"
                width="20"
              >
                <svg>
                  ark.svg
                </svg>
              </div>
            </div>
            <div
              class="relative flex flex-1 h-full"
            >
              <input
                aria-autocomplete="list"
                aria-controls="ImportWallet__network-menu"
                aria-labelledby="ImportWallet__network-label"
                autocomplete="off"
                class="Input__InputStyled-sc-1i8sd1r-1 hPntwa p-0 border-none bg-transparent focus:ring-0 no-ligatures w-full text-theme-secondary-text"
                data-testid="SelectNetworkInput__input"
                disabled=""
                id="ImportWallet__network-input"
                placeholder="Enter a cryptoasset name"
                value="ARK Devnet"
              />
              
            </div>
          </div>
        </div>
        <div
          class=""
          data-testid="SelectNetwork__options"
        >
          <div
            class="mt-6 hidden"
          >
            <ul
              aria-labelledby="ImportWallet__network-label"
              class="grid grid-cols-7 gap-3 mt-3"
              id="ImportWallet__network-menu"
              role="listbox"
            />
          </div>
          <div
            class="mt-6 hidden"
          >
            <ul
              aria-labelledby="ImportWallet__network-label"
              class="grid grid-cols-7 gap-3 mt-3"
              id="ImportWallet__network-menu"
              role="listbox"
            />
          </div>
        </div>
      </div>
    </fieldset>
    <div
      class="flex overflow-hidden flex-col space-y-4 rounded-lg p-4 bg-theme-danger-50 dark:bg-theme-secondary-800 sm:space-y-0 sm:space-x-4 sm:flex-row sm:items-center"
    >
      <div
        class="h-11 w-11 flex flex-shrink-0 items-center justify-center rounded-lg text-white bg-theme-danger-400"
      >
        <div
          class="sc-bdvvtL qGXhc"
          height="20"
          width="20"
        >
          <svg>
            circle-cross.svg
          </svg>
        </div>
      </div>
      <div
        class="flex flex-col w-full"
      >
        <div
          class="text-sm leading-relaxed break-words dark:text-theme-secondary-500"
        >
          Unable to connect to Ledger device. Please ensure that all other applications that connect to your Ledger are closed.
        </div>
      </div>
    </div>
  </section>
</div>
`;

exports[`LedgerConnectionStep should show update error if app version is less than minimum version 1`] = `
<div>
  <section
    class="space-y-8"
    data-testid="LedgerConnectionStep"
  >
    <div
      class="flex justify-between items-end bg-theme-background"
    >
      <div
        class="space-y-4"
      >
        <h1
          class="mb-0 text-2xl"
          data-testid="header__title"
        >
          Open App on Ledger
        </h1>
        <div
          class="text-theme-secondary-text"
          data-testid="header__subtitle"
        >
          Open the app on your Ledger and check for details.
        </div>
      </div>
    </div>
    <fieldset
      class="FormField__FormFieldStyled-sc-q4bblx-0 dFNJMk flex flex-col space-y-2"
    >
      <label
        class="flex mb-2 text-sm font-semibold transition-colors duration-100 FormLabel text-theme-secondary-text"
        data-testid="FormLabel"
        for="network"
      >
        Cryptoasset
      </label>
      <div>
        <div
          aria-expanded="false"
          aria-haspopup="listbox"
          aria-owns="ImportWallet__network-menu"
          data-testid="SelectNetwork"
          role="combobox"
        >
          <label
            for="ImportWallet__network-input"
            id="ImportWallet__network-label"
          />
          
          <div
            class="Input__InputWrapperStyled-sc-1i8sd1r-0 ktoJbG"
            disabled=""
          >
            <div
              aria-label="ARK Devnet"
              class="Circle__CircleWrapper-sc-d5u41q-0 bcplCx border-theme-secondary-200 dark:border-theme-secondary-700 text-theme-secondary-500 ring-theme-background"
              data-testid="SelectNetworkInput__network"
            >
              <div
                class="sc-bdvvtL qGXhc"
                data-testid="NetworkIcon__icon"
                height="20"
                width="20"
              >
                <svg>
                  ark.svg
                </svg>
              </div>
            </div>
            <div
              class="relative flex flex-1 h-full"
            >
              <input
                aria-autocomplete="list"
                aria-controls="ImportWallet__network-menu"
                aria-labelledby="ImportWallet__network-label"
                autocomplete="off"
                class="Input__InputStyled-sc-1i8sd1r-1 hPntwa p-0 border-none bg-transparent focus:ring-0 no-ligatures w-full text-theme-secondary-text"
                data-testid="SelectNetworkInput__input"
                disabled=""
                id="ImportWallet__network-input"
                placeholder="Enter a cryptoasset name"
                value="ARK Devnet"
              />
              
            </div>
          </div>
        </div>
        <div
          class=""
          data-testid="SelectNetwork__options"
        >
          <div
            class="mt-6 hidden"
          >
            <ul
              aria-labelledby="ImportWallet__network-label"
              class="grid grid-cols-7 gap-3 mt-3"
              id="ImportWallet__network-menu"
              role="listbox"
            />
          </div>
          <div
            class="mt-6 hidden"
          >
            <ul
              aria-labelledby="ImportWallet__network-label"
              class="grid grid-cols-7 gap-3 mt-3"
              id="ImportWallet__network-menu"
              role="listbox"
            />
          </div>
        </div>
      </div>
    </fieldset>
    <div
      class="flex overflow-hidden flex-col space-y-4 rounded-lg p-4 bg-theme-danger-50 dark:bg-theme-secondary-800 sm:space-y-0 sm:space-x-4 sm:flex-row sm:items-center"
    >
      <div
        class="h-11 w-11 flex flex-shrink-0 items-center justify-center rounded-lg text-white bg-theme-danger-400"
      >
        <div
          class="sc-bdvvtL qGXhc"
          height="20"
          width="20"
        >
          <svg>
            circle-cross.svg
          </svg>
        </div>
      </div>
      <div
        class="flex flex-col w-full"
      >
        <div
          class="text-sm leading-relaxed break-words dark:text-theme-secondary-500"
        >
          The ARK app version is 1.0.1. Please update the ARK app via Ledger Live.
        </div>
      </div>
    </div>
  </section>
</div>
`;
